### PR TITLE
feat: upgrade dockerfiles to be compatible with python 3.8.0 to 3.12.2

### LINKED ISSUE #68: tutor-ecommerce codebase Python 3.12 upgrade
- [x] Ensure tutor-ecommerce codebase is compatible with Python 3.12
- [x] Update Python to 3.12 in Dockerfile for nightly and ensure the image builds as expected. The changes would be merged to nightly. This task would rely on upgrade of ecommerce to Python 3.12

See https://github.com/overhangio/tutor/issues/1004 for the context.

### PR DESCRIPTION
**Pull request by Faraz32123**
close #68

### CODE REVIEW

**Review by DawoudSheraz — APPROVED** [review]

File: tutorecommerce/templates/ecommerce/build/ecommerce-worker/Dockerfile [hunk 1]
```diff
@@ -15,14 +16,25 @@ RUN mkdir /openedx/ecommerce_worker && \
...
+# Install pyenv
+# https://www.python.org/downloads/
+# https://github.com/pyenv/pyenv/releases
+ARG PYTHON_VERSION=3.12.2
+ENV PYENV_ROOT /opt/pyenv
+USER root
+RUN git clone https://github.com/pyenv/pyenv $PYENV_ROOT --branch v2.3.36 --depth 1
+# Install Python
+RUN $PYENV_ROOT/bin/pyenv install $PYTHON_VERSION
+USER app
```
Comment (DawoudSheraz) [inline tutorecommerce/templates/ecommerce/build/ecommerce-worker/Dockerfile:29]: why do we need to switch users here?

Comment (Faraz32123) [reply tutorecommerce/templates/ecommerce/build/ecommerce-worker/Dockerfile:29]:
```
 => [minimal  7/19] RUN echo "{}" > /openedx/config.yml                                                                          0.2s
 => ERROR [minimal  8/19] RUN git clone https://github.com/pyenv/pyenv /opt/pyenv --branch v2.3.36 --depth 1                     0.4s
------                                                                                                                                
 > importing cache manifest from docker.io/overhangio/openedx-discovery:17.0.0-nightly-cache:
------
------
 > [minimal  8/19] RUN git clone https://github.com/pyenv/pyenv /opt/pyenv --branch v2.3.36 --depth 1:
0.341 fatal: could not create work tree dir '/opt/pyenv': Permission denied
------
Dockerfile:36
--------------------
  34 |     ENV PYENV_ROOT /opt/pyenv
  35 |     # USER root
  36 | >>> RUN git clone https://github.com/pyenv/pyenv $PYENV_ROOT --branch v2.3.36 --depth 1
  37 |     # Install Python
  38 |     RUN $PYENV_ROOT/bin/pyenv install $PYTHON_VERSION
--------------------
ERROR: failed to solve: process "/bin/sh -c git clone https://github.com/pyenv/pyenv $PYENV_ROOT --branch v2.3.36 --depth 1" did not complete successfully: exit code: 128
```

Comment (DawoudSheraz) [reply tutorecommerce/templates/ecommerce/build/ecommerce-worker/Dockerfile:29]: 👍🏽 nit: we can add that as a comment in file that unless switched, we get permission denied. 

But if change user, dont we need to give app user permissions on the repo cloned by root?

Comment (Faraz32123) [reply tutorecommerce/templates/ecommerce/build/ecommerce-worker/Dockerfile:29]: added a comment.
nope we don't need that.

**Comment by regisb** [comment]
Does this mean that ecommerce is working with python 3.12? It would be absolutely fantastic, because we don't have anyone in the community to work on upgrading this IDA https://discuss.openedx.org/t/these-idas-will-not-be-upgraded-to-python-3-11-by-redwood/12755

**Comment by Faraz32123** [comment]
> Does this mean that ecommerce is working with python 3.12? It would be absolutely fantastic, because we don't have anyone in the community to work on upgrading this IDA https://discuss.openedx.org/t/these-idas-will-not-be-upgraded-to-python-3-11-by-redwood/12755

Yes, ecommerce is working using these PRs, [PR for ecommerce](https://github.com/openedx/ecommerce/pull/4147) and [PR for ecommerce-worker](https://github.com/openedx/ecommerce-worker/pull/247).
But still we need to update github workflows for ecommerce's PR as the tests are failing.

**Comment by Faraz32123** [comment]
moved this PR to blocked as this PR cannot be merged until [PR for ecommerce](https://github.com/openedx/ecommerce/pull/4147) and [PR for ecommerce-worker](https://github.com/openedx/ecommerce-worker/pull/247) didn't merge.

**Pull request merged**